Add -lightwalletd experimental option

oxarbitrage opened the pull request:
Similar to `-insightexplorer` but loading less indexes.

After testing and code review this should be able to close https://github.com/zcash/zcash/issues/4326

str4d commented on src/main.cpp:
```diff
@@ -70,7 +70,7 @@ int nScriptCheckThreads = 0;
 std::atomic_bool fReindex(false);
 bool fTxIndex = false;
-bool fAddressIndex = false;     // insightexplorer
+bool fAddressIndex = false;     // insightexplorer && lightwalletd
```
```suggestion
bool fAddressIndex = false;     // insightexplorer || lightwalletd
```

str4d commented on src/main.cpp:
```diff
@@ -4477,6 +4477,15 @@ bool static LoadBlockIndexDB()
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
+    }
```
If both `-insightexplorer` and `-lightwalletd` are set, this causes `-insightexplorer` to be ignored. We should either make them mutually exclusive (only one can be set, enforced by `init.cpp`), or handle the case here where both are set. I'm weakly leaning towards making them mutually exclusive.

daira replied:
Can't this be solved by just setting `fAddressIndex = true;` here? It's not logically inconsistent to want to support both the Insight Explorer and lightwalletd from the same zcashd.

oxarbitrage replied:
Went with @daira suggestion at https://github.com/zcash/zcash/pull/4402/commits/7aed10f195c961408e6c4bbbae2595c2c17e5a89

str4d commented on src/main.cpp:
```diff
@@ -4821,6 +4830,14 @@ bool InitBlockIndex(const CChainParams& chainparams)
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
+    }
```
Ditto.

str4d commented on src/rpc/blockchain.cpp:
```diff
@@ -408,8 +408,8 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
     std::string disabledMsg = "";
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getblockdeltas", "insightexplorer");
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

str4d commented on src/rpc/blockchain.cpp:
```diff
@@ -462,7 +462,7 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
         );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

str4d commented on src/rpc/blockchain.cpp:
```diff
@@ -491,8 +491,8 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
     std::string disabledMsg = "";
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getblockhashes", "insightexplorer");
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```

str4d commented on src/rpc/blockchain.cpp:
```diff
@@ -525,7 +525,7 @@ UniValue getblockhashes(const UniValue& params, bool fHelp)
             );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!fExperimentalInsightExplorer && !fExperimentalLightWalletd) {
```
```suggestion
    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```
and ditto below.

str4d commented on src/rpc/server.cpp:
```diff
@@ -467,14 +467,43 @@ std::string HelpExampleRpc(const std::string& methodname, const std::string& arg
-        "to the zcash.conf file:\n\n"
-        "experimentalfeatures=1\n"
-        + enableArg + "=1\n";
+string experimentalDisabledHelpMsg(const string& rpc, const vector<string>& enableArgs)
```
While we are changing this line, make these:
```suggestion
std::string experimentalDisabledHelpMsg(const std::string& rpc, const std::vector<string>& enableArgs)
```

oxarbitrage commented:
Thanks for the fast and positive review @str4d .

I addressed all the review comments in https://github.com/zcash/zcash/pull/4402/commits/0a194bd6f6a273645ed9a84896dfd8fd65dc8e36 , we can force push after a new review of the changes.

daira commented on qa/rpc-tests/addressindex.py:
```diff
@@ -291,6 +307,15 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
+            assert_equal(deltas[i]['address'],  addr1)
+            assert_equal(deltas[i]['height'],   expected_deltas[i]['height'])
+            assert_equal(deltas[i]['satoshis'], expected_deltas[i]['satoshis'])
+            assert_equal(deltas[i]['txid'],     expected_deltas[i]['txid'])
```
To avoid code duplication, I suggest:
```
        # test getaddressdeltas
        for node in (1, 3):
            deltas = self.nodes[node].getaddressdeltas({'addresses': [addr1]})
            assert_equal(len(deltas), len(expected_deltas))
            for i in range(len(deltas)):
                assert_equal(deltas[i]['address'],  addr1)
                assert_equal(deltas[i]['height'],   expected_deltas[i]['height'])
                assert_equal(deltas[i]['satoshis'], expected_deltas[i]['satoshis'])
                assert_equal(deltas[i]['txid'],     expected_deltas[i]['txid'])
```

daira commented on qa/rpc-tests/addressindex.py:
```diff
@@ -255,6 +266,11 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
+
+        # test getaddressbalance for the lightwalletd node
+        bal = self.nodes[3].getaddressbalance(change)
+        assert(bal['received'] > 0)
```
```suggestion
        # test getaddressbalance
        for node in (2, 3):
            bal = self.nodes[3].getaddressbalance(change)
            assert(bal['received'] > 0)
```

daira commented on src/main.cpp:
```diff
@@ -4468,14 +4468,24 @@ bool static LoadBlockIndexDB()
+    else if (fLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
```
```suggestion
```

daira commented on src/init.cpp:
```diff
@@ -1418,6 +1418,20 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+                    strLoadError = _("Not possible to enable -insightexplorer and -lightwalletd at the same time");
+                    break;
+                }
+
```
```suggestion
```
(requires my other suggested changes below)

daira replied:
If this is not taken, then `s/-lightWalletd/-lightwalletd/` in the comment.

daira commented on src/main.cpp:
```diff
@@ -4815,11 +4825,19 @@ bool InitBlockIndex(const CChainParams& chainparams)
+    else if (fExperimentalLightWalletd) {
+        fAddressIndex = true;
+        fSpentIndex = false;
+        fTimestampIndex = false;
```
```suggestion
```

daira commented on src/rpc/server.cpp:
```diff
@@ -467,14 +467,43 @@ std::string HelpExampleRpc(const std::string& methodname, const std::string& arg
+        "To enable it, restart zcashd with the following command line options:" + newline()
+        + cmd + newline(2) +
+        "Alternatively add these two lines to the zcash.conf file:" + newline(2)
+        + config;
```
```suggestion
    for (size_t i = 0; i < size; ++i)
    {
        if (size == 1 || i == 0)
        {
            cmd += "-experimentalfeatures and -" + enableArgs.at(i);

            config += "experimentalfeatures=1\n";
            config += enableArgs.at(i) + "=1\n";
        }
        else {
            cmd += " or:\n-experimentalfeatures and -" + enableArgs.at(i);

            config += "\nor:\n\n";
            config += "experimentalfeatures=1\n";
            config += enableArgs.at(i) + "=1\n";
        }
    }
    return "\nWARNING: " + rpc + " is disabled.\n" +
        "To enable it, restart zcashd with the following command line options:\n"
        + cmd + "\n\n" +
        "Alternatively add these two lines to the zcash.conf file:\n\n"
        + config;
```

oxarbitrage replied:
Yes, i had it like this before, then added the lambda to clarify a bit the code but actually was not a very good idea. Removed at https://github.com/zcash/zcash/pull/4402/commits/5d3ff01d6177e5fe3974b4483920d794f7e37f12

str4d commented on src/rpc/misc.cpp:
```diff
@@ -1063,8 +1063,8 @@ UniValue getaddresstxids(const UniValue& params, bool fHelp)
-    if (!fExperimentalInsightExplorer) {
-        disabledMsg = experimentalDisabledHelpMsg("getspentinfo", "insightexplorer");
+    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
+        disabledMsg = experimentalDisabledHelpMsg("getspentinfo", {"insightexplorer", "lightwalletd"});
```
Revert these changes; `getspentinfo` cannot be called with `-lightwalletd` because this RPC requires `fSpentIndex` which `-lightwalletd` does not enable.

oxarbitrage replied:
Right, fixed at https://github.com/zcash/zcash/pull/4402/commits/f65acd0aeea44cc4d3ab058ed8b870fb3ba5b595

str4d commented on src/rpc/misc.cpp:
```diff
@@ -1087,7 +1087,7 @@ UniValue getspentinfo(const UniValue& params, bool fHelp)
         );
 
-    if (!fExperimentalInsightExplorer) {
+    if (!(fExperimentalInsightExplorer || fExperimentalLightWalletd)) {
```
Ditto.

str4d commented on qa/rpc-tests/addressindex.py:
```diff
@@ -255,6 +266,12 @@ def check_balance(node_index, address, expected_balance, expected_received=None)
         change = change_vout[0]['scriptPubKey']['addresses'][0]
         bal = self.nodes[2].getaddressbalance(change)
         assert(bal['received'] > 0)
```
Remove these two lines (they are checked immediately below).

oxarbitrage commented:
Removed duplicated lines from test, squashed everything with the first commit by `git revise` and force pushed.

LarryRuane commented:
> ... (without rebasing on master, so GitHub's force-push UI shows no change) ...

@str4d, what is the GitHub force-push UI?

str4d commented:
@LarryRuane when you force-push to a branch that has a PR, GitHub adds a note to the PR's timeline. Clicking on the words "force-pushed" shows the changes across the force-push. [In this case](https://github.com/zcash/zcash/compare/f65acd0aeea44cc4d3ab058ed8b870fb3ba5b595..4d0b20882cb4ff728901018828526584131709d7), the only changes were the ones that addressed my comment.

str4d commented:
Nice work!

@zkbot r+

oxarbitrage commented:
Failed by pyflakes: `qa/rpc-tests/addressindex.py:19: 'test_framework.util.start_node' imported but unused`.

Fixed and force pushed, please retry.

str4d commented:
Confirmed that the force-push [only adjusted the test imports](https://github.com/zcash/zcash/compare/4d0b20882cb4ff728901018828526584131709d7..a43f5daff8819520da32ea7569033625b8ca656f).

@zkbot r+

str4d commented:
An unexpected transient failure; opened #4438.

@zkbot retry

str4d commented:
More transient failures...

@zkbot retry
